upload-notebooks/upload-notebook.py

```diff
@@ -0,0 +1,120 @@
+    description="Upload notebook to TileDB Cloud",
+    epilog="TileDB Cloud options may also be set as environment variables, but values set via command line flags will take priority",
+)
+parser.add_argument(
+    "notebook",
+    help="Notebook file to upload (extension .ipynb)",
+)
+parser.add_argument(
+    "--tiledb-cloud-token",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_TOKEN"),
+    help="API token for TileDB Cloud account",
+)
+parser.add_argument(
+    "--tiledb-cloud-namespace",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_NAMESPACE"),
+    help="Namespace for TileDB Cloud",
+)
+parser.add_argument(
+    "--tiledb-cloud-storage-path",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_STORAGE_PATH"),
+    help="Storage path on TileDB Cloud",
+)
+parser.add_argument(
+    "--tiledb-cloud-storage-credential-name",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_STORAGE_CREDENTIAL_NAME"),
+    help="Storage credentials to use on TileDB Cloud",
+)
+args = parser.parse_args()
+
+notebook = args.notebook
+token = args.tiledb_cloud_token
+namespace = args.tiledb_cloud_namespace
+storage_path = args.tiledb_cloud_storage_path
+storage_credential_name = args.tiledb_cloud_storage_credential_name
+
+# Verify input
+if not os.path.exists(notebook):
+    sys.stderr.write("Error: Notebook file does not exist: %s\n" % (notebook))
+    sys.exit(1)
+
+notebook_name, ext = os.path.splitext(notebook)
+notebook_name = os.path.basename(notebook_name)
+if ext != ".ipynb":
+    sys.stderr.write(
+        "Error: Notebook file extension must be '.ipynb', not '%s'\n" % (ext)
+    )
+    sys.exit(1)
+
+if token is None:
+    sys.stderr.write("Error: Missing API token\n")
+    sys.stderr.write(
+        "You must define the token as the env var TILEDB_CLOUD_TOKEN or pass it to --tiledb-cloud-token\n"
+    )
+    sys.exit(1)
+
+if namespace is None:
+    sys.stderr.write("Error: Missing namespace\n")
+    sys.stderr.write(
+        "You must define the namespace as the env var TILEDB_CLOUD_NAMESPACE or pass it to --tiledb-cloud-namespace\n"
+    )
+    sys.exit(1)
+
+# Login and upload
+tiledb.cloud.login(token=token)
+user = tiledb.cloud.user_profile()
+sys.stderr.write("Info: Logged into TileDB Cloud as %s\n" % (user.username))
+sys.stderr.write("Info: Notebook to upload is %s\n" % (notebook))
+sys.stderr.write("Info: Namespace is %s\n" % (namespace))
+if storage_path is not None:
+    sys.stderr.write("Info: Storage path is %s\n" % (storage_path))
+if storage_credential_name is not None:
+    sys.stderr.write(
+        "Info: Storage credential name is %s\n" % (storage_credential_name)
+    )
+
+try:
+    tiledb.cloud.upload_notebook_from_file(
+        ipynb_file_name=notebook,
+        namespace=namespace,
+        array_name=notebook_name,
+        storage_path=storage_path,
+        storage_credential_name=storage_credential_name,
+        on_exists=tiledb.cloud.notebook.OnExists.OVERWRITE,
```

Comment:
@thetorpedodog `OnExists.OVERWRITE` _sounds_ like it should work even if the file doesn't exist? Is there a bug here or can we clarify this in the future?

Reply:
At least in practice, I wasn't able to upload a new notebook when using `OnExists.OVERWRITE`, that's why I switched to using `try`/`except` ([commit](https://github.com/jdblischak/TileDB-BioImaging/commit/0aeaffc307c08a86d003d0554da3dd8b6c8b93ab)). Here's the full traceback ([link to GitHub Actions log](https://github.com/jdblischak/TileDB-BioImaging/actions/runs/4365951816/jobs/7635337205#step:4:28)):

```python
Traceback (most recent call last):
  File "/home/runner/work/_temp/bcd054aa-18c6-4e42-b7dd-8b2a8e9c6f14.py", line 6, in <module>
    tiledb.cloud.upload_notebook_from_file(
  File "/home/runner/.local/lib/python3.10/site-packages/tiledb/cloud/notebook.py", line 152, in upload_notebook_from_file
    return upload_notebook_contents(
  File "/home/runner/.local/lib/python3.10/site-packages/tiledb/cloud/notebook.py", line 215, in upload_notebook_contents
    _write_notebook_to_array(tiledb_uri, ipynb_file_contents, ctx)
  File "/home/runner/.local/lib/python3.10/site-packages/tiledb/cloud/notebook.py", line 356, in _write_notebook_to_array
    with tiledb.open(tiledb_uri, mode="w", ctx=ctx) as arr:
  File "/home/runner/.local/lib/python3.10/site-packages/tiledb/highlevel.py", line 21, in open
    return tiledb.Array.load_typed(
  File "tiledb/libtiledb.pyx", line 1859, in tiledb.libtiledb.Array.load_typed
  File "tiledb/libtiledb.pyx", line 1682, in tiledb.libtiledb.preload_array
  File "tiledb/libtiledb.pyx", line 451, in tiledb.libtiledb._raise_ctx_err
  File "tiledb/libtiledb.pyx", line 436, in tiledb.libtiledb._raise_tiledb_error
tiledb.cc.TileDBError: [TileDB::Array] Error: [TileDB::REST] Error: Error in libcurl GET operation: libcurl error message 'CURLE_OK'; HTTP code 404; server response data '{"code":604,"message":"Unrecognized array","request_id":"f0a324ff-8d26-49bc-9c77-bc2c412de21a"}'.
```

Reply:
It's been a while since I was in this code but it does seem like `OnExists.OVERWRITE` is *intended* to work if the notebook does not already exist, and if it doesn’t, that would be a bug on some level.

Reply:
Ok, thanks for confirming. I can file a bug report in the [TileDB-Cloud-Py](https://github.com/TileDB-Inc/TileDB-Cloud-Py) repo